library/src/main/java/com/novoda/downloadmanager/LiteDownloadMigrationService.java

```diff
@@ -81,65 +35,91 @@ public void onCreate() {
-        }
-
+        notificationChannelCreator = new MigrationNotificationChannelCreator(getResources());
         notificationCreator = new MigrationNotification(getApplicationContext(), android.R.drawable.ic_dialog_alert);
+        notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
 
         super.onCreate();
     }
 
+    class MigrationDownloadServiceBinder extends Binder {
+        DownloadMigrationService getService() {
+            return LiteDownloadMigrationService.this;
+        }
+    }
+
     @Nullable
     @Override
     public IBinder onBind(Intent intent) {
         return binder;
     }
 
     @Override
-    public void onTaskRemoved(Intent rootIntent) {
-        Log.d(TAG, "onTaskRemoved");
-        rescheduleMigration();
-        Log.d(TAG, "Rescheduling");
-        super.onTaskRemoved(rootIntent);
+    public void setMigrationCallback(MigrationServiceBinder.Callback migrationCallback) {
+        this.migrationCallback = migrationCallback;
     }
 
-    private void rescheduleMigration() {
-        Intent intent = new Intent(getApplicationContext(), LiteDownloadMigrationService.class);
-        PendingIntent pendingIntent = PendingIntent.getService(this, 1, intent, PendingIntent.FLAG_ONE_SHOT);
-        AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
-        if (alarmManager == null) {
-            Log.w(TAG, "Could not retrieve AlarmManager for rescheduling.");
-            return;
-        }
-        alarmManager.set(AlarmManager.RTC_WAKEUP, SystemClock.elapsedRealtime() + 5000, pendingIntent);
+    @Override
+    public void setNotificationChannelCreator(NotificationChannelCreator notificationChannelCreator) {
+        this.notificationChannelCreator = notificationChannelCreator;
     }
 
     @Override
-    public boolean onUnbind(Intent intent) {
-        Log.d(TAG, "onUnbind");
-        migrationCallback = null;
-        Log.d(TAG, "Stopping service");
-        return super.onUnbind(intent);
+    public void setNotificationCreator(NotificationCreator<MigrationStatus> notificationCreator) {
+        this.notificationCreator = notificationCreator;
     }
 
-    class MigrationDownloadServiceBinder extends Binder {
-
-        MigrationDownloadServiceBinder withCallback(Migrator.Callback migrationCallback) {
-            LiteDownloadMigrationService.this.migrationCallback = migrationCallback;
-            return this;
+    @Override
+    public void startMigration() {
+        Optional<NotificationChannel> notificationChannel = notificationChannelCreator.createNotificationChannel();
+        String channelName = notificationChannelCreator.getNotificationChannelName();
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && notificationChannel.isPresent() && notificationChannelDoesNotExist(channelName)) {
+            notificationManager.createNotificationChannel(notificationChannel.get());
         }
```

Comment:
For clarity I think we can move the version and channel existence to a separate method. Feels like these 3 lines are doing too much.
```
if(notificationChannel.isPresent()) {
  createNotificationChannel(notificationChannel.get())
}
```
It could even be part of a "compat" utils module/library